src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -340,10 +338,26 @@ void FileSegment::setRemoteFileReader(RemoteFileReaderPtr remote_file_reader_)
     remote_file_reader = remote_file_reader_;
 }
 
-void FileSegment::write(char * from, size_t size, size_t offset)
+void FileSegment::write(char * from, size_t size, size_t offset_in_file)
 {
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::FileSegmentWriteMicroseconds);
+    assertWriteAllowed(size, offset_in_file);
+    writeImpl(size, offset_in_file, [&]()
```

Comment:
Should be there another `write` overload witch uses `writeImpl`? It looks like out opted to changing internal `cache_writer`, so can this change with passing lambda be undone?